### DISCUSSION TITLE
Make 3D model transparent by setting alpha values, change blending mode at runtime.

### DISCUSSION THREAD

**Discussion by MukeshBlippar** [sceneview/sceneview#200] [Ideas]
Hi, I'm using your library and it's wonderful. I have requirement to set Fade in / out animation on 3D models. And also to change the opacity of any model at run time to make it transparent. So far from all the issues in sceneform and filament, I learnt that it's only possible if the blending mode of model is set 'Transparent' or 'Fade'. But the default blending mode is 'Opaque' which is not allowing to set the alpha values. And Filament API also doesn't allow to change the blending mode at run time.
I managed to set opacity for the model by creating my own custom material and applying the on renderable. But this over rides the existing material. So the question is how can we achieve to set the alpha value in existing material. Any plan to add this feature in future? Here is my code which is working  to create material with alpha value. 

```
 Material materialFilament;
 MaterialBuilder.init();
        MaterialBuilder materialBuilder = new MaterialBuilder()
                .platform(MaterialBuilder.Platform.MOBILE)
                .name("External Glb Material")
                .require(MaterialBuilder.VertexAttribute.UV0)
                .shading(MaterialBuilder.Shading.UNLIT)
                .doubleSided(false)
                .optimization(MaterialBuilder.Optimization.NONE);

        MaterialPackage materialPackage = materialBuilder
                .blending(MaterialBuilder.BlendingMode.TRANSPARENT)
                .material("void material(inout MaterialInputs material) {\n" +
                        "    prepareMaterial(material);\n" +
                        "    material.baseColor.a = 0.30;\n" +
                        "    material.baseColor.rgb *= material.baseColor.a;\n" +
                        "}\n")
                .build(filamentEngine);

        if (materialPackage.isValid()) {
            ByteBuffer buffer = materialPackage.getBuffer();
            materialFilament = new com.google.android.filament.Material.Builder()
                    .payload(buffer, buffer.remaining())
                    .build(filamentEngine);
        }
        MaterialBuilder.shutdown();
```

**Comment by ThomasGorisse** [comment]
Hi.
In your case, I would strongly advice having your model directly comming from a gltf (You can easily create your shapes using Blender and export as glTF).
Once your model is loaded, you will just have to change it material parameters without having to apply a new one from scratch:
```kotlin
node.modelInstance?.materialInstance?.setParameter(
          "baseColorFactor",
          // Red
          0.09f,
          // Green
          0.44f,
          // Blue
          0.95f,
          // Alpha
          0.75f
      )
```

**Reply by MukeshBlippar** [reply]
Hi @ThomasGorisse  Thanks for your quick reply. Above solution is only working if the blending mode of the model is already set 'transparent/alpha blend'. Is there any way to change the blending mode using scene form/filament on runtime inside the app. Our models have opaque blending by default.

**Reply by ThomasGorisse** [reply]
You should ask it on the Filament repository

**Reply by MukeshBlippar** [reply]
Yes, I will ask there also. I was just wondering if there is some plan to add such feature at scene from api level. That's why I added in discussion here. Would be great if fadein/out animation you can add.

**Comment by ThomasGorisse** [comment]
And change your model `Blend Mode` to `Alpha Blend` if you want it to handle transparency.
<img width="309" alt="image" src="https://user-images.githubusercontent.com/6597529/211664244-9d432b8d-e99b-481d-ac7d-b1f7e57049c1.png">

**Comment by ThomasGorisse** [comment]
In v4.0.0-rc.1, you can adjust transparency via `MaterialInstance` parameters:

```kotlin
val modelInstance = rememberModelInstance(modelLoader, "models/mymodel.glb")
if (modelInstance != null) {
    val material = modelInstance.materialInstances.first()
    // Set base color with alpha (RGBA)
    material.setParameter("baseColorFactor", 1f, 1f, 1f, 0.5f)
}
```

For the blending mode to work, the material in your GLB must be set to "BLEND" mode (not "OPAQUE"). You can configure this in your 3D authoring tool (Blender: set material blend mode to Alpha Blend) or via Filament material definitions.

**Comment by Daniel-Shegida** [comment]
Can anyone see comments and replies? For me it is 3 comments and replies. Is there a solution to this? I also want to change transparency of models with opaque material but i am stuck at copying material too.
